heat/core/types.py

```diff
@@ -854,6 +861,87 @@ def promote_types(
     return __type_promotions[typecode_type1][typecode_type2]
 
 
+def result_type(
+    *arrays_and_types: Tuple[Union[dndarray.DNDarray, Type[datatype], Any]]
+) -> Type[datatype]:
+    """
+    Returns the data type that results from type promotions rules performed in an arithmetic operation.
+
+    Parameters
+    ----------
+    arrays_and_types: List of arrays and types
+        Input arrays, types or numbers of the operation.
+
+    Examples
+    --------
+    >>> ht.result_type(ht.array([1], dtype=ht.int32), 1)
+    ht.int32
+    >>> ht.result_type(ht.float32, ht.array(1, dtype=ht.int8))
+    ht.float32
+    >>> ht.result_type("i8", "f4")
+    ht.float64
+    """
+
+    def result_type_rec(*arrays_and_types):
+        # derive type
+        arg = arrays_and_types[0]
+
+        try:
+            # array / tensor
+            if isinstance(arg, np.ndarray):
+                type1 = canonical_heat_type(arg.dtype.char)
+            else:
+                type1 = canonical_heat_type(arg.dtype)
+
+            if len(arg.shape) > 0:
+                prec1 = 0
+            else:
+                prec1 = 2
+        except Exception:
+            try:
+                # type
+                if isinstance(arg, np.dtype):
+                    arg = arg.char
+                type1 = canonical_heat_type(arg)
+                prec1 = 1
+            except Exception:
+                # instance
+                type1 = canonical_heat_type(type(arg))
+                prec1 = 3
+
+        # multiple arguments
+        if len(arrays_and_types) > 1:
+            type2, prec2 = result_type_rec(*arrays_and_types[1:])
+
+            if type1 == type2:
+                return type1, min(prec1, prec2)
+            else:
+                if prec1 == prec2:
+                    return promote_types(type1, type2), prec1
+
+                for sclass in (bool, integer, float):
+                    if issubdtype(type1, sclass) and issubdtype(type2, sclass):
+                        if prec1 < prec2:
+                            return type1, min(prec1, prec2)
+                        if prec1 > prec2:
+                            return type2, min(prec1, prec2)
+
+                tc1 = __type_codes[type1]
+                tc2 = __type_codes[type2]
+
+                if tc1 < tc2:
+                    return type2, min(prec1, prec2)
+                if tc1 > tc2:
+                    return type1, min(prec1, prec2)
+
+                raise RuntimeError("Something went wrong")  # pragma: no cover
```

Comment:
can drop these down some tabs, the if loops end with returns

Reply:
Yes, I flattened the else block.